feat: add C `ndarray` and refactor for `stats/base/dmeanpn`

DhruvArvindSingh opened the pull request:
Resolves None.

## Description

> What is the purpose of this pull request?

This pull request:

-   adds C `ndarray` implementation for `stats/base/dmeanpn`
-   Updates README.md

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves None. 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

DhruvArvindSingh committed:
feat: add C ndarray implementation for stats/base/dmeanpn

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

DhruvArvindSingh committed:
feat: add C ndarray implementation for stats/base/dmeanpn

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: passed
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: passed
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

DhruvArvindSingh commented:
/stdlib update-copyright-years

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/src/main.c:
```diff
@@ -35,29 +36,48 @@
+}
+
+/**
+* Computes the arithmetic mean of a double-precision floating-point strided array using a two-pass error correction algorithm.
```
This is the incorrect JSDOC api call it should be :- Computes the arithmetic mean of a double-precision floating-point strided array using a two-pass error correction algorithm and alternative indexing semantics.

DhruvArvindSingh replied:
Thanks for pointing out ... I have made the necessary changes

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/include/stdlib/stats/base/dmeanpn.h:
```diff
@@ -31,7 +31,13 @@ extern "C" {
+double API_SUFFIX( stdlib_strided_dmeanpn )( const CBLAS_INT N, const double *X, const CBLAS_INT strideX );
+
+/**
+* Computes the arithmetic mean of a double-precision floating-point strided array using a two-pass error correction algorithm.
```
Same here

DhruvArvindSingh replied:
Thanks for pointing out ... I have made the necessary changes

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/src/main.c:
```diff
@@ -35,29 +36,48 @@
 *
 * @param N       number of indexed elements
 * @param X       input array
```
also might want to stack these output types one above the another!

DhruvArvindSingh replied:
Thanks for pointing out ... I have made the necessary changes

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/lib/ndarray.js:
```diff
@@ -54,21 +54,21 @@ var dapxsumpw = require( '@stdlib/blas/ext/base/dapxsumpw' ).ndarray;
 * var v = dmeanpn( N, x, 2, 1 );
```
N should be replaced 4 here

Neerajpathak07 replied:
here and the native file as well

DhruvArvindSingh replied:
Thanks for pointing out ... I have made the necessary changes

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/lib/dmeanpn.native.js:
```diff
@@ -40,10 +40,10 @@ var addon = require( './../src/addon.node' );
 * var N = x.length;
 *
 * var v = dmeanpn( N, x, 1 );
```
N should be replaced by x.length here

Neerajpathak07 replied:
in the `dmeanpn.js` file as well

DhruvArvindSingh replied:
Thanks for pointing out ... I have made the necessary changes

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/lib/dmeanpn.js:
```diff
@@ -52,23 +52,9 @@ var dapxsumpw = require( '@stdlib/blas/ext/base/dapxsumpw' );
 * var v = dmeanpn( N, x, 1 );
```
same here

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/src/addon.c:
```diff
@@ -17,6 +17,7 @@
 */
 
 #include "stdlib/stats/base/dmeanpn.h"
+#include "stdlib/blas/base/shared.h"
```
Not sure if this is supposed to be added here

Neerajpathak07 replied:
Oh alright, I must have missed that!
Pls do look into my other comments as well

DhruvArvindSingh replied:
Thanks for pointing out ... I have made the necessary changes

Neerajpathak07 commented:
@DhruvArvindSingh 
Hey, left a bunch of comments do look into it also you might need to add some changes in all the test files replacing N with 4 and removing N = floor/2 
The `examples.c` will also require changes setting a const array and replacing `stride` with `strideX`

Neerajpathak07 commented:
I would suggest you to make the suggested changes in your other PR's as well.
It makes it difficult for the maintainers while reviewing the code.
Had similar issues with my PR's as well which is why thought about bringing this up.

DhruvArvindSingh commented:
Thanks for pointing out ... I have made the necessary changes

DhruvArvindSingh committed:
Merge branch 'stdlib-js:develop' into feat/ndarray_dmeanpn

DhruvArvindSingh committed:
fix: fixed minor error

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: passed
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

Neerajpathak07 commented:
@DhruvArvindSingh 
Apparently changes made by you have not been committed here looks like they underwent merge branch commit .
As I'm  unable to view the applied changes!!

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/lib/index.js:
```diff
@@ -28,9 +28,8 @@
-* var N = x.length;
 *
-* var v = dmeanpn( N, x, 1 );
+* var v = dmeanpn( 3, x, 1 );
```
This should be x.length

DhruvArvindSingh replied:
I don't think its compulsary to put `x.length` in some files while absolute in others

Neerajpathak07 replied:
if you would see the previous set value of N was x.length which why on removing it the function should also call the value set by N
To answer your query absolute values are given in repl and markdown files when we have a defined array.

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/test/test.dmeanpn.native.js:
```diff
@@ -177,7 +177,7 @@ tape( 'if provided an `N` parameter equal to `1`, the function returns the first
 	t.end();
 });
 
-tape( 'the function supports a `stride` parameter', opts, function test( t ) {
```
remove :-
```javascript
N = floor( x.length / 2 );
```
and change var v to:-
```javascript
v = dmeanpn( 4, x, 2 );
```

Neerajpathak07 replied:
similarly make these changes through out all the other test files and for the suitable test cases also

Neerajpathak07 replied:
remove:-
```javascript
var floor = require( '@stdlib/math/base/special/floor' );
```
as you won't be needing it in the test case

DhruvArvindSingh replied:
Resolved

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/lib/dmeanpn.native.js:
```diff
@@ -37,9 +37,12 @@ var addon = require( './../src/addon.node' );
+<<<<<<< HEAD
 * var v = dmeanpn( N, x, 1 );
+=======
+* var v = dmeanpn( 3, x, 1 );
```
same should be x.length not 3

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/lib/dmeanpn.js:
```diff
@@ -47,9 +47,8 @@ var ndarray = require( './ndarray.js' );
-* var N = x.length;
 *
-* var v = dmeanpn( N, x, 1 );
+* var v = dmeanpn( 3, x, 1 );
```
same comment

DhruvArvindSingh replied:
I don't think its compulsary to put x.length in some files while absolute in others

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/examples/c/example.c:
```diff
@@ -28,10 +28,10 @@ int main( void ) {
 	int64_t N = 4;
```
in the example.c file make the following changes:-
```c
const double x[] = { 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0 };

const int N = 4;

const int strideX = 2
```

Neerajpathak07 replied:
remove:- 
```#include <stdint.h>```

Neerajpathak07 commented:
You will also need to add changes to `repl.txt` and `index.d.ts` for instance you can refer this PR:- #4251 and how the refactoring is done over there.

DhruvArvindSingh committed:
fix: fixed test files

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/README.md:
```diff
@@ -69,9 +69,9 @@ The function has the following parameters:
 -   **N**: number of indexed elements.
 -   **x**: input [`Float64Array`][@stdlib/array/float64].
--   **stride**: index increment for `x`.
+-   **strideX**: index increment for `x`.
```
```suggestion
-   **strideX**: stride length for `x`.
```

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/README.md:
````diff
@@ -101,7 +101,7 @@ var v = dmeanpn( N, x1, 2 );
 ```
 
-#### dmeanpn.ndarray( N, x, stride, offset )
+#### dmeanpn.ndarray( N, x, stridXe, offset )
````
```suggestion
#### dmeanpn.ndarray( N, x, strideX, offsetX )
```

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/README.md:
````diff
@@ -175,6 +175,107 @@ console.log( v );
+```c
+const double x[] = { 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 };
+
+double v = stdlib_strided_dmeankbn2_ndarray( 4, x, 2, 1 );
````
There is no need for calling  `dmeankb2` when you are implementing `dmeanpn`

Neerajpathak07 replied:
```suggestion
double v = stdlib_strided_dmeanpn_ndarray( 4, x, 2, 1 );
```

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/src/main.c:
```diff
@@ -19,45 +19,47 @@
 * @param X       input array
-* @param stride  stride length
+* @param strideX  strideX length
 * @return        output value
```
```suggestion
* @return         output value
```

DhruvArvindSingh committed:
chore: minor issues

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: passed
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

DhruvArvindSingh committed:
refactor: refactored benchmark

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

DhruvArvindSingh committed:
Merge branch 'stdlib-js:develop' into feat/ndarray_dmeanpn

DhruvArvindSingh committed:
fix: src/main.c

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/lib/ndarray.js:
```diff
@@ -49,26 +49,25 @@ var dapxsumpw = require( '@stdlib/blas/ext/base/dapxsumpw' ).ndarray;
 * var floor = require( '@stdlib/math/base/special/floor' );
```
```suggestion

```

Neerajpathak07 replied:
do this for  `ndarray.native.js` file as well.

Neerajpathak07 replied:
just a last change then this should be up.

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/lib/ndarray.js:
```diff
@@ -49,26 +49,25 @@ var dapxsumpw = require( '@stdlib/blas/ext/base/dapxsumpw' ).ndarray;
 * // returns 1.25
 */
-function dmeanpn( N, x, stride, offset ) {
+function dmeanpn( N, x, strideX, offset ) {
```
```suggestion
function dmeanpn( N, x, strideX, offsetX ) {
```

Neerajpathak07 replied:
change every offset element to offsetX

DhruvArvindSingh replied:
Resolved

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/test/test.dmeanpn.js:
```diff
@@ -138,7 +133,7 @@ tape( 'if provided a `stride` parameter equal to `0`, the function returns the f
 
 	x = new Float64Array( [ 1.0, -2.0, -4.0, 5.0, 3.0 ] );
 
-	v = dmeanpn( x.length, x, 0 );
```
there is no need to remove this revert it back

DhruvArvindSingh replied:
Resolved

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/test/test.dmeanpn.native.js:
```diff
@@ -216,14 +212,13 @@ tape( 'the function supports a negative `stride` parameter', opts, function test
 });
 
-tape( 'if provided a `stride` parameter equal to `0`, the function returns the first element', opts, function test( t ) {
+tape( 'if provided a `strideX` parameter equal to `0`, the function returns the first element', opts, function test( t ) {
```
```suggestion
tape( 'if provided a `stride` parameter equal to `0`, the function returns the first element', opts, function test( t ) {
```
there is no need to change the description of the test case.

Neerajpathak07 replied:
change it for all the test files

DhruvArvindSingh replied:
Resolved

Neerajpathak07 commented:
@DhruvArvindSingh This PR is getting in line now just a few comments here and there.

DhruvArvindSingh committed:
Merge branch 'stdlib-js:develop' into feat/ndarray_dmeanpn

DhruvArvindSingh committed:
fix: tests of stats/base/dmeanpn

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

DhruvArvindSingh committed:
Update lib/node_modules/@stdlib/stats/base/dmeanpn/lib/ndarray.js

Co-authored-by: Neeraj Pathak <neerajrpathak710@gmail.com>
Signed-off-by: Dhruv/ <154677013+DhruvArvindSingh@users.noreply.github.com>

DhruvArvindSingh commented:
Thanks @Neerajpathak07 for pointing out my mistakes.

Neerajpathak07 commented:
@DhruvArvindSingh 
My pleasure! and thanks for bearing with me XD.

DhruvArvindSingh commented:
@Neerajpathak07  since you have approved my PR you have to comment `lgtm` on this PR inorder to approve it

Neerajpathak07 commented:
@DhruvArvindSingh 
Hey, I didn't have time to have a look at the the entire PR which is why I didn't mention it in the first place. Will take a look once I get free from college!!

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/README.md:
````diff
@@ -175,6 +168,101 @@ console.log( v );
+Computes the [arithmetic mean][arithmetic-mean] of a double-precision floating-point strided array using aa two-pass error correction algorithm.
+
+```c
+const double x[] = { 1.0, -2.0, 2.0 };
````
```suggestion
const double x[] = { 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0 };
```

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/README.md:
````diff
@@ -175,6 +168,101 @@ console.log( v );
+```c
+const double x[] = { 1.0, -2.0, 2.0 };
+
+double v = stdlib_strided_dmeanpn( x.length, x, 1 );
````
```suggestion
double v = stdlib_strided_dmeanpn( 4, x, 2 );
```

Neerajpathak07 replied:
this change still needs to be addressed as we don't pass `x.length` in the `stdlib_strided_dmeanpn()`

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/README.md:
```diff
@@ -175,6 +168,101 @@ console.log( v );
+const double x[] = { 1.0, -2.0, 2.0 };
+
+double v = stdlib_strided_dmeanpn( x.length, x, 1 );
+// returns ~0.3333
```
```suggestion
// returns ~1.25
```

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/README.md:
```diff
@@ -175,6 +168,101 @@ console.log( v );
+int main( void ) {
+    // Create a strided array:
+    const double x[] = { 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0 };
+
```
```suggestion

// Specify the number of elements:
    const int N = 4;
    
```

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/benchmark/benchmark.js:
```diff
@@ -39,13 +45,7 @@ var dmeanpn = require( './../lib/dmeanpn.js' );
-	for ( i = 0; i < x.length; i++ ) {
-		x[ i ] = ( randu()*20.0 ) - 10.0;
-	}
+	var x = uniform( len, -10.0, 10.0, options );
```
Need to use `filledarrayBy` and a custom random function.

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/benchmark/benchmark.js:
```diff
@@ -21,14 +21,20 @@
+
+
 // FUNCTIONS //
 
```
for Ex:-
```javascript
/**
* Returns a random value or `NaN`.
*
* @private
* @returns {number} random number or `NaN`
**/
function rand() {
	if ( bernoulli( 0.2 ) ) {
		return NaN;
	}
	return uniform( -10.0, 10.0 );
}
```

Neerajpathak07 replied:
Same goes for the rest of the `benchmark` files

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/examples/c/example.c:
```diff
@@ -17,21 +17,14 @@
-
-	// Specify the stride length:
-	int64_t stride = 2;
+	const double x[] = { 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0 };
```
```suggestion
	const double x[] = { 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0 };
	
	// Specify the number of elements:
	const int N = 4;

	// Specify the stride length:
	const int strideX = 2;
```

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/examples/c/example.c:
```diff
@@ -17,21 +17,14 @@
 
 	// Compute the arithmetic mean:
-	double v = stdlib_strided_dmeanpn( N, x, stride );
+	double v = stdlib_strided_dmeanpn( 4, x, 2 );
```
```suggestion
	double v = stdlib_strided_dmeanpn( N, x, strideX );
```

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/examples/index.js:
```diff
@@ -18,18 +18,12 @@
-}
+var x = uniform( 10, -10.0, 10.0, {
+	'dtype': 'float64'
+});
```
Revert the code for the `index.js` back to the original code.
As you won't be needing a `uniform` function after you are done adding a custom random function.

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/lib/dmeanpn.js:
```diff
@@ -40,35 +40,20 @@ var dapxsumpw = require( '@stdlib/blas/ext/base/dapxsumpw' );
-
-	return mu + c;
+function dmeanpn( N, x, strideX ) {
+	var ox = stride2offset( N, strideX );
```
```suggestion

```

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/lib/dmeanpn.js:
```diff
@@ -40,35 +40,20 @@ var dapxsumpw = require( '@stdlib/blas/ext/base/dapxsumpw' );
-	return mu + c;
+function dmeanpn( N, x, strideX ) {
+	var ox = stride2offset( N, strideX );
+	return ndarray( N, x, strideX, ox );
```
```suggestion
	return ndarray( N, x, strideX, stride2offset( N, strideX ) );
```

Neerajpathak07 replied:
this is just my opinion for this as it directly passes the result of `stride2offset(N, strideX)` as an argument to `ndarray`

DhruvArvindSingh committed:
fix: fixing mistakes

Co-authored-by: Neeraj Pathak <neerajrpathak710@gmail.com>
Signed-off-by: Dhruv/ <154677013+DhruvArvindSingh@users.noreply.github.com>

DhruvArvindSingh committed:
fix: Update README.md

Signed-off-by: Dhruv/ <154677013+DhruvArvindSingh@users.noreply.github.com>

DhruvArvindSingh committed:
fix: Update README.md

Signed-off-by: Dhruv/ <154677013+DhruvArvindSingh@users.noreply.github.com>

DhruvArvindSingh committed:
fix: Update example.c

Signed-off-by: Dhruv/ <154677013+DhruvArvindSingh@users.noreply.github.com>

DhruvArvindSingh committed:
fix: update example.c

Signed-off-by: Dhruv/ <154677013+DhruvArvindSingh@users.noreply.github.com>

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/README.md:
```diff
@@ -175,6 +168,104 @@ console.log( v );
+    const int N = 4;
+    
+    // Compute the arithmetic mean:
+    double v = stdlib_strided_dmeanpn( N, x, 2 );
```
```suggestion
    double v = stdlib_strided_dmeanpn( N, x, strideX );
```

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/README.md:
```diff
@@ -175,6 +168,104 @@ console.log( v );
+    // Specify the number of elements:
+    const int N = 4;
+    
+    // Compute the arithmetic mean:
```
```suggestion
    // Specify the stride length:
    const int strideX = 2;
    
    // Compute the arithmetic mean:
```

Neerajpathak07 commented:
@DhruvArvindSingh 
Just a few comments other than that everything else looks good!!

Neerajpathak07 commented:
Also the changes done in the index.js should also be done in the markdown file.

Neerajpathak07 commented:
@DhruvArvindSingh Would you mind taking a look at this PR will be needing this for refactoring `dmean`

Neerajpathak07 commented:
Hey @DhruvArvindSingh Hope you're doing great!
Could you a look at this PR will be needing this for refactoring `dmean`.

DhruvArvindSingh commented:
> Hey @DhruvArvindSingh Hope you're doing great! Could you a look at this PR will be needing this for refactoring `dmean`.

Sorry for the late reply, i was actually busy with my exams but now i am free..... I will resolve this by tomorrow morning

Neerajpathak07 commented:
Great, you'll also need to update it to handle `NaN` elements you can refer to some of the merged Pr's or for instance take a look at this #4496

DhruvArvindSingh committed:
Update lib/node_modules/@stdlib/stats/base/dmeanpn/README.md

Co-authored-by: Neeraj Pathak <neerajrpathak710@gmail.com>
Signed-off-by: Dhruv/ <154677013+DhruvArvindSingh@users.noreply.github.com>

DhruvArvindSingh committed:
Update lib/node_modules/@stdlib/stats/base/dmeanpn/README.md

Co-authored-by: Neeraj Pathak <neerajrpathak710@gmail.com>
Signed-off-by: Dhruv/ <154677013+DhruvArvindSingh@users.noreply.github.com>

ShabiShett07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/README.md:
```diff
@@ -175,6 +168,107 @@ console.log( v );
+    const double x[] = { 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0 };
+
+    // Specify the number of elements:
+    const int N = 4;
+    
+    // Specify the stride length:
```
```suggestion
    const int N = 4;

    // Specify the stride length:
```

ShabiShett07 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/README.md:
```diff
@@ -175,6 +168,107 @@ console.log( v );
+    const int N = 4;
+    
+    // Specify the stride length:
+    const int strideX = 2;
+    
+    // Compute the arithmetic mean:
```
```suggestion
    const int strideX = 2;

    // Compute the arithmetic mean:
```
